### PR TITLE
fix(core): adjust for graphemes width when calculating word boundaries

### PR DESCRIPTION
**Pull request by CruelAddict**
PR on issue #566 

Added a helper calculating column delta and applied it to getPrevWordBoundary and getNextWordBoundary. The algorithm is basically reused from segment_callback

Disclaimer: I don't really know zig and this is my first time using it :)

### CODE REVIEW

**Comment by kommander** [comment]
Currently `findWrapBreaks` does not resolve col offsets, that is right. Having it do that is very much overhead. My plan here was:

- let `getGraphemes` do grapheme and word break parsing in one go for a TextChunk, as it iterates anyway and can then cache both.
- `getGraphemes` is currently triggered in word wrapping logic anyway
- word breaks can then have col offsets, as the grapheme parsing iterates graphemes and col offsets anyway

This should overall be more efficient as currently doing graphemes and word breaks separately (two loops) and would make additional conversion/translation obsolete. 

cc @simonklee wdyt?

**Comment by simonklee** [comment]
> cc @simonklee wdyt?

I think the perf regression is localized to getNextWordBoundary/getPrevWordBoundary, not the main render/wrap loop. From that perspective it should be okay. 

Maybe add tests for emoji/CJK and tabs in next/prev word boundary? Something like.

```diff
diff --git i/packages/core/src/edit-buffer.test.ts w/packages/core/src/edit-buffer.test.ts
index 1a54e1a..d36776d 100644
--- i/packages/core/src/edit-buffer.test.ts
+++ w/packages/core/src/edit-buffer.test.ts
@@ -417,6 +417,18 @@ describe("EditBuffer", () => {
       const next1 = buffer.getNextWordBoundary()
       expect(next1.col).toBeGreaterThan(0)
     })
+
+    it("should handle wide graphemes before boundaries", () => {
+      buffer.setText("你 好")
+      buffer.setCursorToLineCol(0, 0)
+
+      const nextBoundary = buffer.getNextWordBoundary()
+      expect(nextBoundary.col).toBe(3)
+
+      buffer.setCursorToLineCol(0, 5)
+      const prevBoundary = buffer.getPrevWordBoundary()
+      expect(prevBoundary.col).toBe(3)
+    })
   })
 
   describe("native coordinate conversion methods", () => {
diff --git i/packages/core/src/zig/tests/edit-buffer_test.zig w/packages/core/src/zig/tests/edit-buffer_test.zig
index 4b2ab74..09c33b9 100644
--- i/packages/core/src/zig/tests/edit-buffer_test.zig
+++ w/packages/core/src/zig/tests/edit-buffer_test.zig
@@ -120,6 +120,38 @@ test "EditBuffer - multiple word boundaries" {
     try std.testing.expectEqual(@as(u32, 16), cursor.col);
 }
 
+test "EditBuffer - next word boundary with wide grapheme" {
+    const pool = gp.initGlobalPool(std.testing.allocator);
+    defer gp.deinitGlobalPool();
+
+    var eb = try EditBuffer.init(std.testing.allocator, pool, .wcwidth);
+    defer eb.deinit();
+
+    // "你" = 2 cols, " " = 1 col, "好" = 2 cols
+    try eb.insertText("你 好");
+    try eb.setCursor(0, 0);
+
+    const next_cursor = eb.getNextWordBoundary();
+    try std.testing.expectEqual(@as(u32, 0), next_cursor.row);
+    try std.testing.expectEqual(@as(u32, 3), next_cursor.col);
+}
+
+test "EditBuffer - prev word boundary with wide grapheme" {
+    const pool = gp.initGlobalPool(std.testing.allocator);
+    defer gp.deinitGlobalPool();
+
+    var eb = try EditBuffer.init(std.testing.allocator, pool, .wcwidth);
+    defer eb.deinit();
+
+    // "你" = 2 cols, " " = 1 col, "好" = 2 cols
+    try eb.insertText("你 好");
+    try eb.setCursor(0, 5);
+
+    const prev_cursor = eb.getPrevWordBoundary();
+    try std.testing.expectEqual(@as(u32, 0), prev_cursor.row);
+    try std.testing.expectEqual(@as(u32, 3), prev_cursor.col);
+}
+
 test "EditBuffer - word boundary at end of line" {
     const pool = gp.initGlobalPool(std.testing.allocator);
     defer gp.deinitGlobalPool();
```

**Comment by CruelAddict** [comment]
Added tests as requested.
I've also found a problem with line width calculation in segment callback (fix [here](https://github.com/anomalyco/opentui/pull/567/files#diff-ca9b30db0c25eba0062f2a40d0b01a7ce0540a5c48a5e94a5317efd6ee50244bR1137))
Recorded a little repro:
https://github.com/user-attachments/assets/012a55ae-c3c7-4af0-90bf-48df30a73d6b

And added a test catching the issue. It also felt right to add similar tests on emoji & CJK there too

**Comment by kommander** [comment]
Ah cool, I've been looking for that CJK line break bug. Github has some issue running the core build and test job currently it seems. Will try again in a bit.
